risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+/// The irreducible polynomial `x^2 - 11` was chosen because `11` is
+/// the simplest choice of `B` for `x^2 - B` that makes this polynomial
+/// irreducible.
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Pod, Zeroable)]
+#[repr(transparent)]
+pub struct ExtElem([Elem; EXT_SIZE]);
+
+impl Default for ExtElem {
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+impl field::Elem for ExtElem {
+    const ZERO: ExtElem = ExtElem::zero();
+    const ONE: ExtElem = ExtElem::one();
+
+    /// Generate a random [ExtElem] uniformly.
+    fn random(rng: &mut impl rand::Rng) -> Self {
+        Self([Elem::random(rng), Elem::random(rng)])
+    }
+
+    /// Raise an [ExtElem] to a power of `n`.
+    fn pow(self, n: usize) -> Self {
+        let mut n = n;
+        let mut tot = ExtElem::from(1);
+        let mut x = self;
+        while n != 0 {
+            if n % 2 == 1 {
+                tot *= x;
+            }
+            n = n / 2;
+            x *= x;
+        }
+        tot
+    }
+
+    /// Compute the multiplicative inverse of a field element [ExtElem].
+    fn inv(self) -> Self {
+        let a = &self.0;
+        // Compute the multiplicative inverse by looking at Fp2 as a composite field and
+        // using the same basic methods used to invert complex numbers.
+        // Starting with field element `a`, we begin with the initial value `out = 1 /
+        // a`. Setting `a'` to be `a` with a_1 negated, we multiply the
+        // numerator and the denominator by `a'` to produce `a' / (a * a') =
+        // (a_0 - a_1) / (a_0^2 + B * a_1^2)`. To safely compute this value,
+        // we multiply by the safe inverse of the denominator.
+        let det: Elem = a[0] * a[0] + BETA * a[1] * a[1];
+        let invdet: Elem = det.inv();
+        ExtElem([a[0] * invdet, -a[1] * invdet])
+    }
+}
+
+impl field::ExtElem for ExtElem {
+    const EXT_SIZE: usize = EXT_SIZE;
+
+    type SubElem = Elem;
+
+    fn from_subfield(elem: &Elem) -> Self {
+        Self::from([elem.clone(), Elem::ZERO])
+    }
+}
+
+impl From<[Elem; EXT_SIZE]> for ExtElem {
+    /// Create field element from subfield element
+    fn from(val: [Elem; EXT_SIZE]) -> Self {
+        ExtElem(val)
+    }
+}
+
+const BETA: Elem = Elem::new(11u64);
+const NBETA: Elem = Elem::new(P - 11);
+
+impl ExtElem {
+    /// Explicitly construct a field element [ExtElem] from parts.
+    pub fn new(x0: Elem, x1: Elem) -> Self {
+        Self([x0, x1])
+    }
+
+    /// Create a [ExtElem] from a [Elem].
+    pub fn from_fp(x: Elem) -> Self {
+        Self([x, Elem::new(0)])
+    }
+
+    /// Create a [ExtElem] from a raw integer.
+    pub const fn from_u64(x0: u64) -> Self {
+        Self([Elem::new(x0), Elem::new(0)])
+    }
+
+    /// Returns the value zero.
+    const fn zero() -> Self {
+        Self::from_u64(0)
+    }
+
+    /// Returns the value one.
+    const fn one() -> Self {
+        Self::from_u64(1)
+    }
+
+    /// Returns the constant portion of a [Elem].
```

Comment:
IMO "constant portion" is unclear. The algorithm here is that given an element of the form `a + b * sqrt(BETA)`, return `a`. Maybe say "base field portion"?... open to other suggestions...